File: six/HistogramLine.java
package six;
/* HistogramLine.java 
 * Created on Oct 23, 2006 
 * by ilteris
 */
import processing.core.PApplet;

public class HistogramLine {
	PApplet	p;
	Vector3D	loc, vel, acc;
	Vector3D  destLoc;
	
	private float maxvel;
	private float	damper;

	private float damp,mass,k;
	public HistogramLine(PApplet $p, Vector3D loc_) {
		p = $p;
		loc = loc_;
		destLoc = loc.copy();
		vel = new Vector3D(0.0f, 0, 0);
		acc = new Vector3D(0.0f, 0, 0);
		maxvel = 7;
		damper = p.random(0.2f,0.8f);
		
		damp = 0.9f; 
		mass = 15; 
		k = 1;

	}

	

	public void update() {
		vel = Vector3D.sub(destLoc, loc);
		vel.mult(damper);
		vel.mult(0.3f);
		loc.add(vel);
		
		vel.limit(maxvel);
		 
	}

	public void spring() {
		
		    Vector3D force = Vector3D.sub(loc,destLoc);
		    force.mult(-k);
		    force.div(mass);

		    acc.setXY(0,0);
		    acc.add(force);
		    vel.add(acc);
		    vel.mult(damp);
		    loc.add(vel);

		  
	}

	public void setDestLoc(Vector3D destLoc_) {
		
		destLoc = destLoc_;
	}

}

File: HistogramLine.js
class HistogramLine {
    constructor(p5Instance, loc_) {
        this.p = p5Instance;
        this.loc = loc_.copy();
        this.destLoc = loc_.copy();
        this.vel = this.p.createVector(0, 0, 0); // p5.Vector
        this.acc = this.p.createVector(0, 0, 0); // p5.Vector
        this.maxvel = 7;
        this.damper = this.p.random(0.2, 0.8);
        this.damp = 0.9;
        this.mass = 15;
        this.k = 1;
    }

    spring() {
        let force = p5.Vector.sub(this.loc, this.destLoc);
        force.mult(-this.k);
        force.div(this.mass);
        this.acc.set(0, 0); // Assuming 2D for now
        this.acc.add(force);
        this.vel.add(this.acc);
        this.vel.mult(this.damp);
        this.loc.add(this.vel);
    }

    setDestLoc(destLoc_) {
        this.destLoc = destLoc_.copy();
    }

    update() {
        this.vel = p5.Vector.sub(this.destLoc, this.loc);
        this.vel.mult(this.damper);
        this.vel.mult(0.3); // This was originally mult(0.03) but seems too small, using 0.3 as in Island.java's direct use
        this.loc.add(this.vel);
        this.vel.limit(this.maxvel); // In Processing, limit was on PVector, p5.js vel.limit is correct
    }

    // Added for drawing, not in original spec but useful for testing
    display() {
        this.p.ellipse(this.loc.x, this.loc.y, 5, 5);
    }
}
